File: Mbs/Trading/Data/Generators/RepetitiveSample/RepetitiveSampleGeneratorParameters.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;

namespace Mbs.Trading.Data.Generators.RepetitiveSample
{
    /// <summary>
    /// The input parameters for the repetitive sample generator.
    /// </summary>
    public class RepetitiveSampleGeneratorParameters : IValidatableObject
    {
        /// <summary>
        /// Gets or sets the number of samples to generate.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [Range(2, int.MaxValue, ErrorMessage = ErrorMessages.FieldMustBePositiveMinValue2)]
        public int SampleCount { get; set; } = DefaultParameterValues.SamplesCount;

        /// <summary>
        /// Gets or sets the time related input parameters.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [ValidateObject]
        public TimeParameters TimeParameters { get; set; } = new TimeParameters();

        /// <summary>
        /// Gets or sets the number of samples before the very first waveform sample.
        /// The value of zero means the waveform starts immediately.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.OffsetSamples)]
        [Range(0, int.MaxValue, ErrorMessage = ErrorMessages.FieldMustNotBeNegative)]
        public int OffsetSamples { get; set; } = DefaultParameterValues.OffsetSamples;

        /// <summary>
        /// Gets or sets the number of repetitions of the waveform. The value of zero means infinite.
        /// </summary>
        [Required(ErrorMessage = ErrorMessages.FieldIsRequired, AllowEmptyStrings = false)]
        [DefaultValue(DefaultParameterValues.RepetitionsCount)]
        [Range(0, int.MaxValue, ErrorMessage = ErrorMessages.FieldMustNotBeNegative)]
        public int RepetitionsCount { get; set; } = DefaultParameterValues.RepetitionsCount;

        /// <inheritdoc />
        public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
        {
            yield break;
        }
    }
}

File: Mbs.Api.Host.Ng/ClientApp/src/app/shared/mbs/data/generators/repetitive-sample/repetitive-sample-generator-parameters.ts
import { TimeParameters } from '../time-parameters';
import { sampleCountName, offsetSamplesName, repetitionsCountName, timeParametersName, objectName } from '../constants';

/** The input parameters for the repetitive sample generator. */
export class RepetitiveSampleGeneratorParameters {
  private static readonly defaultSampleCount: number = 252;
  private static readonly defaultOffsetSamples: number = 0;
  private static readonly defaultRepetitionsCount: number = 0;

  /** The number of samples to generate. */
  sampleCount: number = RepetitiveSampleGeneratorParameters.defaultSampleCount;

  /** The number of samples before the very first waveform sample.
   *
   * The value of zero means the waveform starts immediately. */
  offsetSamples: number = RepetitiveSampleGeneratorParameters.defaultOffsetSamples;

  /** The number of repetitions of the waveform.
   *
   * The value of zero means infinite. */
  repetitionsCount: number = RepetitiveSampleGeneratorParameters.defaultRepetitionsCount;

  /** The time related input parameters. */
  timeParameters: TimeParameters = new TimeParameters();

  constructor(data?: RepetitiveSampleGeneratorParameters) {
    if (data) {
      for (const property in data) {
        if (data.hasOwnProperty(property)) {
          (<any>this)[property] = (<any>data)[property];
        }
      }
    }
  }

  toJSON(data?: any) {
    data = typeof data === objectName ? data : {};
    data[sampleCountName] = this.sampleCount;
    data[offsetSamplesName] = this.offsetSamples;
    data[repetitionsCountName] = this.repetitionsCount;
    data[timeParametersName] = this.timeParameters ? this.timeParameters.toJSON() : new TimeParameters();
    return data;
  }
}
